autorel/clients/http_client.py

```diff
@@ -0,0 +1,173 @@
+            - Some of them takes None value for now
+        """
+        if self.DEBUG_FLAG:  # Debug only when the debug flag is set
+            new_frame = [reqHeaders, None, None, None]
+            if self._frameCount < self.DEBUG_FRAME_BUFFER_SIZE-1:
+                self._frameBuffer.append(new_frame)
+            else:
+                self._frameCount = -1
+                self._frameBuffer[0] = new_frame
+            self._frameCount += 1
+
+    def DEBUG_ON_RESPONSE(self, respHeader, respStatus, respData):
+        """
+            Add the API response to the debug frame
+        """
+        if self.DEBUG_FLAG:  # Debug only when the debug flag is set
+            self._frameBuffer[self._frameCount][1:4] = [respHeader, respStatus, respData]
+            self._headers[self.DEBUG_HEADER_KEY] = self._frameCount
+
+    def DEBUG_LOGGING(self):
+        if self.DEBUG_LOGGING:
+            # log the frames to the terminal, pass for now
+            pass
+
+    def carryRequest(self, method="GET", url="", headers={},
+                     params={}, payload={}):
+        """
+        Carry out a {method} request.
+        - The url can be a FQDN( Fully qualified Domain name ) but should have netloc same as that
+          of API endpoint.
+        """
+        method = method.lower()
+        try:
+            methodToCall = getattr(requests, method)
+        except:
+            e = Exception("The method {0} is not supported".format(method))
+            raise ClientException(e)
+        request_url = url if (url.startswith("http") or
+            url.startswith('https')) else urljoin(self._endpoint, url)
+        headers.update(self._headers)
+        params.update(self._params)
+        timeout = self._timeout
+        self.NEW_DEBUG_FRAME(headers)
+        try:
+            if method == "post" or method == "put":
+                # encode the payload to json if required
+                if headers["Content-Type"] == "application/json":
+                    data = json.dumps(payload)
+                # the payload will be automatically form-encoded elsewise
+                else:
+                    data = payload
+                resp = methodToCall(request_url, headers=headers,
+                                    params=params, data=data,
+                                    timeout=timeout, verify=False
+                                    )
+            else:
+                resp = methodToCall(request_url, headers=headers,
+                                    params=params, data=payload,
+                                    timeout=timeout
+                                    )
+            # Debug on response
+            self.DEBUG_ON_RESPONSE(resp.headers, resp.status_code, resp.json())
+            # Github API always returns a JSON (https://developer.github.com/v3/#schema)
+            # resp.json() is safe, it won't throw a JSONDecodingError
+            return (resp.status_code, resp.headers, resp.json())
```

Comment:
I think it's quite common but @lbudai may have different thoughts. But as far as a proper response is considered I can pass the requests response i.e `resp` but that would require `GithubInterface` to do accessing which is not good considering the abstraction we are adding here.